cli/cli/helpers/grafloki/kubernetes_grafloki.go

```diff
@@ -0,0 +1,477 @@
+				},
+				ResizePolicy:             nil,
+				VolumeDevices:            nil,
+				LivenessProbe:            nil,
+				ReadinessProbe:           nil,
+				StartupProbe:             nil,
+				Lifecycle:                nil,
+				TerminationMessagePath:   "",
+				TerminationMessagePolicy: "",
+				ImagePullPolicy:          "",
+				SecurityContext:          nil,
+				Stdin:                    false,
+				StdinOnce:                false,
+				TTY:                      false,
+			},
+		},
+		[]apiv1.Volume{{
+			Name: "datasources",
+			VolumeSource: apiv1.VolumeSource{
+				ConfigMap: &apiv1.ConfigMapVolumeSource{
+					LocalObjectReference: apiv1.LocalObjectReference{
+						Name: grafanaDatasourceConfigMapName,
+					},
+					Items:       nil,
+					DefaultMode: nil,
+					Optional:    nil,
+				},
+				HostPath:              nil,
+				EmptyDir:              nil,
+				GCEPersistentDisk:     nil,
+				AWSElasticBlockStore:  nil,
+				GitRepo:               nil,
+				Secret:                nil,
+				NFS:                   nil,
+				ISCSI:                 nil,
+				Glusterfs:             nil,
+				PersistentVolumeClaim: nil,
+				RBD:                   nil,
+				FlexVolume:            nil,
+				Cinder:                nil,
+				CephFS:                nil,
+				Flocker:               nil,
+				DownwardAPI:           nil,
+				FC:                    nil,
+				AzureFile:             nil,
+				VsphereVolume:         nil,
+				Quobyte:               nil,
+				AzureDisk:             nil,
+				PhotonPersistentDisk:  nil,
+				Projected:             nil,
+				PortworxVolume:        nil,
+				ScaleIO:               nil,
+				StorageOS:             nil,
+				CSI:                   nil,
+				Ephemeral:             nil,
+			},
+		}},
+		&apiv1.Affinity{
+			NodeAffinity:    nil,
+			PodAffinity:     nil,
+			PodAntiAffinity: nil,
+		})
+	shouldRemoveGrafanaDeployment := true
+	defer func() {
+		if shouldRemoveGrafanaDeployment {
+			if err := k8sManager.RemoveDeployment(ctx, graflokiNamespace, grafanaDeployment); err != nil {
+				logrus.Warnf("Attempted to remove Loki deployment after an error occurred creating it but an error occurred removing it.")
+				logrus.Warnf("Manually remove Loki deployment with name: %v", lokiDeployment.Name)
+			}
+		}
+	}()
+	if err != nil {
+		return "", stacktrace.Propagate(err, "An error occurred creating Grafana deployment.")
+	}
+
+	grafanaService, err := k8sManager.CreateService(ctx,
+		graflokiNamespace,
+		grafanaServiceName,
+		map[string]string{}, // empty labels
+		nil,                 // empty annotations
+		grafanaLabels,       // match grafana deployment pod labels
+		apiv1.ServiceTypeNodePort,
+		[]apiv1.ServicePort{{
+			Name:        "grafana-dashboard",
+			Port:        grafanaNodePort,
+			TargetPort:  intstr.FromInt(grafanaPort),
+			Protocol:    apiv1.ProtocolTCP,
+			NodePort:    grafanaNodePort,
+			AppProtocol: &httpApplicationProtocol,
+		}})
+	shouldRemoveGrafanaService := true
+	defer func() {
```

Comment:
Defer-undo before err-checking, logtext description, loudness of the user action, etc.

Will skip commenting on this further since I think you get the idea lol